libvast/src/format/csv.cpp

```diff
@@ -129,4 +140,271 @@ const char* writer::name() const {
   return "csv-writer";
 }
 
+using namespace parser_literals;
+
+reader::reader(caf::atom_value table_slice_type,
+               std::unique_ptr<std::istream> in)
+  : super(table_slice_type) {
+  if (in != nullptr)
+    reset(std::move(in));
+}
+
+void reader::reset(std::unique_ptr<std::istream> in) {
+  VAST_ASSERT(in != nullptr);
+  input_ = std::move(in);
+  lines_ = std::make_unique<detail::line_range>(*input_);
+}
+
+caf::error reader::schema(vast::schema s) {
+  schema_ = std::move(s);
+  for (auto& t : s) {
+    auto r = caf::get_if<record_type>(&t);
+    if (!r)
+      continue;
+    schema_.add(flatten(*r));
+  }
+  return caf::none;
+}
+
+vast::schema reader::schema() const {
+  return schema_;
+}
+
+const char* reader::name() const {
+  return "csv-reader";
+}
+
+caf::optional<record_type>
+reader::make_layout(const std::vector<std::string>& names) {
+  for (auto& t : schema_) {
+    if (auto r = caf::get_if<record_type>(&t)) {
+      auto select_fields = [&]() -> caf::optional<record_type> {
+        std::vector<record_field> result_raw;
+        for (auto& name : names) {
+          if (auto field = r->at(name))
+            result_raw.emplace_back(name, *field);
+          else
+            return caf::none;
+        }
+        return record_type{result_raw};
+      };
+      if (auto result = select_fields())
+        return result->name(r->name());
+    } else if (names.size() == 1 && names[0] == t.name()) {
+      // Hoist naked type into record.
+      // TODO: Maybe this is actually not a good idea?
+      return record_type{{t.name(), t}}.name(t.name());
+    } // else skip
+  }
+  return caf::none;
+}
+
+namespace {
+
+template <class Iterator, class Attribute>
+struct container_parser_builder {
+  using result_type = rule<Iterator, Attribute>;
+
+  explicit container_parser_builder(const std::string& set_separator)
+    : set_separator_{set_separator} {
+    // nop
+  }
+
+  // TODO: support enumeration_type inside of containers too.
+  template <class T>
+  result_type operator()(const T& t) const {
+    if constexpr (std::is_same_v<T, string_type>) {
+      return +(parsers::any - set_separator_ - '=')->*[](std::string&& x) {
+        return data{std::move(x)};
+      };
+    } else if constexpr (std::is_same_v<T, pattern_type>) {
+      return +(parsers::any - set_separator_ - '=')->*[](std::string&& x) {
+        return data{pattern{std::move(x)}};
+      };
+    } else if constexpr (std::is_same_v<T, set_type>) {
+      auto set_insert = [](std::vector<Attribute>&& xs) {
+        return set(std::make_move_iterator(xs.begin()),
+                   std::make_move_iterator(xs.end()));
+      };
+      return ('{'_p >> (caf::visit(*this, t.value_type) % set_separator_)
+              >> '}'_p)
+               ->*set_insert;
```

Comment:
The line breaking makes this hard to read. Mind refactoring it here (and also in similar occurrences below)?